sdk/core/azure-core/src/main/java/com/azure/core/http/HttpPipelineNextPolicy.java

```diff
@@ -34,16 +46,24 @@ public class HttpPipelineNextPolicy {
      * @return A publisher which upon subscription invokes next policy and emits response from the policy.
      */
     public Mono<HttpResponse> process() {
-        final int size = this.pipeline.getPolicyCount();
-        if (this.currentPolicyIndex > size) {
-            return Mono.error(new IllegalStateException("There is no more policies to execute."));
-        }
-
-        this.currentPolicyIndex++;
-        if (this.currentPolicyIndex == size) {
-            return this.pipeline.getHttpClient().send(this.context.getHttpRequest(), this.context.getContext());
+        if (originatedFromSyncContext && !Schedulers.isInNonBlockingThread()) {
+            // Pipeline executes in synchronous style. We most likely got here via default implementation in the
+            // HttpPipelinePolicy.processSynchronously so go back to sync style here.
+            // Don't do this on non-blocking threads.
+            return Mono.fromCallable(() -> new HttpPipelineNextSyncPolicy(state).processSync());
         } else {
-            return this.pipeline.getPolicy(this.currentPolicyIndex).process(this.context, this);
+            if (originatedFromSyncContext) {
+                LOGGER.warning("The pipeline switched from synchronous to asynchronous."
+                    + " Check if all policies override HttpPipelinePolicy.processSync");
```

Comment:
Can we log the policy name as well here to easily identify which policy caused this switch? The logger only logs the name of this class (HttpPipelineNextPolicy).